File: coci/2014/COCI_14_P5_SABOR.cpp
#include <stdio.h>
#include <string.h>
#include <vector>

#define scan(x) do{while((x=getchar())<'0'); for(x-='0'; '0'<=(_=getchar()); x=(x<<3)+(x<<1)+_-'0');}while(0)
char _;

using namespace std;

vector <bool> arr;
vector <vector <int> > list;
vector <int> inter;

int main () {
	int N; scan (N);
	
	list = vector <vector <int> > (N);
	arr = vector <bool> (N);
	inter = vector <int> (N);
	
	for (int n = 0; n < N; n++) {
		arr [n] = false;
	}
	
	int P, S, E;
	
	for (int d = 0; d < 5; d++) {
		scan (P);
		
		for (int p = 0; p < P; p++) {
			scan (S); scan (E);
			S--; E--;
			
			list [S].push_back (E);
			list [E].push_back (S);
			
			inter [S]++; inter [E]++;
		}
	}
	
	while (true) {
		bool change = false;
		
		for (int c = 0; c < N; c++) {
			if (inter [c] > 2) {
				arr [c] = !arr [c];
				change = true;
				
				for (int i = 0; i < list [c].size (); i++) {
					if (arr [c] == arr [list [c][i]]) {
						inter [c]++;
						inter [list [c][i]]++;
					}
					else {
						inter [c]--;
						inter [list [c][i]]--;
					}
				}
			}
		}
		
		if (!change) {
			break;
		}
	}
	
	for (int n = 0; n < N; n++) {
		printf ("%c", !arr [n] ? 'B' : 'A');
	}
}

File: coci/2014/COCI_14_P5_SABOR.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class COCI_14_P5_SABOR {

	static List <Integer>[] list;
	static char[] arr;
	static int[] inter;

	public static void opp (int i) {
		arr[i] = (arr[i] == 'A' ? 'B' : 'A');
	}

	public static void main (String[] t) throws IOException {
		BufferedReader in = new BufferedReader (new InputStreamReader (System.in));
		int N = Integer.parseInt (in.readLine ()), S, E;
		list = new ArrayList[N];
		arr = new char[N];
		inter = new int[N];

		Arrays.fill (arr, 'B');

		for (int d = 0; d < 5; d++) {
			t = in.readLine ().split (" +");

			//for (String i : t) {
			//  System.out.print ("'" + i + "' ");
			// }

			for (int i = 1; i < t.length / 2 + 1; i++) {
				S = Integer.parseInt (t[i * 2 - 1]) - 1;
				E = Integer.parseInt (t[i * 2]) - 1;

				if (list[S] == null) {
					list[S] = new ArrayList <Integer> ();
				}

				if (list[E] == null) {
					list[E] = new ArrayList <Integer> ();
				}

				list[S].add (E);
				list[E].add (S);

				inter[S]++;
				inter[E]++;
			}
		}

		while (true) {
			boolean change = false;

			for (int c = 0; c < arr.length; c++) {
				if (inter[c] > 2) {
					opp (c);
					change = true;

					if (list[c] != null) {
						for (int i = 0; i < list[c].size (); i++) {
							if (arr[c] == arr[list[c].get (i)]) {
								inter[c]++;
								inter[list[c].get (i)]++;
							}
							else {
								inter[c]--;
								inter[list[c].get (i)]--;
							}
						}
					}
				}
			}

			if (!change) {
				break;
			}
		}

		for (int c = 0; c < arr.length; c++) {
			System.out.print (arr[c]);
		}
	}
}